bpo-42944 Fix Random.sample when counts is not None

jonanifranco opened the pull request:
Use the instance generator instead of the module generator.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-42944](https://bugs.python.org/issue42944) -->
https://bugs.python.org/issue42944
<!-- /issue-number -->

jonanifranco committed:
Fix Random.sample when counts is not None

Replace [Random.]sample by self.sample in Random.sample path when count is not None to correctly use instance generator instead of module generator.

serhiy-storchaka reviewed (commented):
Interesting, that there is a test which should catch this bug, but it does not work because it tests only global `random.sample()`.

Please replace `random` with `self.gen` in `test_sample_counts_equivalence` and move that method to `MersenneTwister_TestBasicOps`, because it does not make sense for SystemRandom.

jonanifranco committed:
Add test for bpo-42944

jonanifranco committed:
Move test_sample_counts_equivalence

test_sample_counts_equivalence should catch bpo-42944, but it does not work because it tests only global random.sample.
random is replaced by self.gen in test_sample_counts_equivalence and that method is moved to MersenneTwister_TestBasicOps, because it does not make sense for SystemRandom

jonanifranco committed:
Fix random.Random.sample when counts is not None

Fidget-Spinner commented on Misc/NEWS.d/next/Library/2021-01-18-10-41-44.bpo-42944.RrONvy.rst:
```diff
@@ -0,0 +1 @@
+Fix `random.Random.sample` when `counts` argument is not `None`.
```
Travis CI/Azure pipelines checks are failing because you need double backticks here ;).
```suggestion
Fix ``random.Random.sample`` when ``counts`` argument is not ``None``.
```

jonanifranco replied:
OK ! Thank you.

jonanifranco committed:
Update Misc/NEWS.d/next/Library/2021-01-18-10-41-44.bpo-42944.RrONvy.rst

Co-authored-by: Ken Jin <28750310+Fidget-Spinner@users.noreply.github.com>
